### ISSUE TITLE
curious type checking error

### ISSUE THREAD

**Issue by maddanio** [microsoft/pyright#1745]
**Describe the bug**
the code below leads to a pyright error where it says that `None` cannot be assigned to `None | str` which is obviously wrong. curiously I could not reduce it further, even the `len(n)` cannot be just replaced by an int.

**To Reproduce**
enter the code below in a pyright-enabled ide

**Expected behavior**
no error

**Screenshots or Code**
```python
from typing import Optional, Dict, List
a: Dict[str, List[Optional[str]]] = dict({
    n: [None] * len(n)
    for n in ['a']
})
```

**VS Code extension or command-line**
pyright-LSP plugin in sublime

**Comment by erictraut** [comment]
This is related to this issue: https://github.com/microsoft/pyright/issues/1691

**Comment by maddanio** [comment]
so the heuristic just doesnt quite cover this case? how can I understand the `Type "None" cannot be assigned to type "int | None"` part of the error message? Just reading it it seems plain wrong (assuming `|` denotes a union type)

**Comment by erictraut** [comment]
The message `Type "None" cannot be assigned to type "int | None"` is part of a larger error message. It's referring to the fact that the type parameter used for the `List` class is invariant, and therefore it must match exactly.

Consider the following:
```python
a: List[None] = [None]
b: List[Optional[int]] = a   # Type error
```

The second assignment is not allowed because `List[None]` is not assignable to type `List[Optional[int]]` (which is the same as `List[int | None]`). This may seem unintuitive at first, but there's a good reason for it. For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/type-concepts.md#generic-types).

Whereas the above code generates a type error, the following similar sample does not.
```python
b: List[Optional[int]] = [None]
```

The reason this doesn't generate an error is because the inferred type of the right-hand side (RHS) of the assignment expression is being influenced by the declared type of the LHS. This is referred to as "bidirectional type inference", and you can read more about it [here](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#bidirectional-type-inference-expected-types).

Pyright supports bidirectional type inference in a bunch of situations, but there are cases that it cannot handle. You've hit one of those situations. In general, bidirectional type inference cannot be applied to expressions that involve binary operators like `*` or `+` or `in`. That's because these operators involve complex operator overloads in Python. For more details, refer to the discussion in #1691. Based on that discussion, I decided to add special-case bidirectional type inference support specifically for the pattern `[<expression>] * <integer literal>`. For example, the following works fine.

```python
b: List[Optional[int]] = [None] * 4
```

But this does not:
```python
x = 4
b: List[Optional[int]] = [None] * x
```

I need to think more about whether I can extend the special-case handling to cover more cases, including the one in your code sample above. Every special case adds complexity to the type checker (and often introduces undesirable edge-case behaviors and bugs), so I try to avoid them where possible.

**Comment by erictraut** [comment]
I've extended the existing special case to handle arbitrary expressions on the RHS of an expression of the form `[expression>] * <RHS>`. Previously, the code handled bidirectional type inference only when RHS was an integer literal. It now handles other expression types including call expressions like in your example.

**Comment by maddanio** [comment]
ok, cool, sounds like it makes sense. On a more general note I am wondering, when assigning containers and the contained types mismatch only by rhs being strictly contained in the lhst union type, would there be any hurt in allowing the assignemt? due to python not really knowing about unions to python those two types are 100% the same and will behave that way, there is no conversion loss or type changes happening.

i.e. assigning `List[a]` to `List[Union[a,b]]` and so on for any rhs union being a strict subset of the lhs union.

**Comment by erictraut** [comment]
If you have a union that includes multiple subtypes, it's fine to assign one of those subtypes to a union. For example, if the LHS is type `Union[A, B]`, you can assign either `A` or `B` to it.

```python
a: A = A()
a: B = B()
y: Union[A, B] = a  # Allowed
z: Union[A, B] = b  # Allowed
```

However, the same does not apply in the case of `List[Union[A, B]]` and `List[A]`. The latter is not a subtype of the former. That's because the type parameter for `List` is "invariant". For more details on invariance (and covariance, contravariance), refer to PEP 483. Also refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/type-concepts.md#generic-types).

```python
a: List[A] = []
y: List[Union[A, B]] = a  # Not allowed
```

This isn't allowed because `List` is a mutable container type. If you were allowed to assign `a` to `y` above, you would then be allowed to push a new `B` onto list `y`, but now `a` (which is the same list) would also contain a `B` element, and that would be a type violation.

You can do what you're suggesting for immutable containers, which use "covariant" type parameters. The immutable form of a `List` is a `Sequence`.

```python
a: List[A] = []
y: List[Union[A, B]] = a  # Not allowed
z: Sequence[Union[A, B]] = a  # Allowed
```

**Comment by maddanio** [comment]
Ok, yes, because immutables behave like values, while all others are basically shared pointers, got it

**Comment by erictraut** [comment]
This is included in Pyright 1.1.130, which I just published. It will also be part of the next release of Pylance.
